### PR TITLE
feat: Migrate to sp1

### PR DESCRIPTION
**Pull request by MaanavKhaitan**
# What

- Migrates foundry template to use sp1
- Adds `build.rs` to build the consumer's zkVM program(s)
- Uses `ivm-zkvm-executor` and `ivm-abi` instead of rewriting the contract ABI, executor functions, etc.
- Changes the type of `programID` from `bytes32` to `bytes`, since we are using `ivm-zkvm-executor` which uses `bytes` now
- Updates `README`

# Testing

`forge test -vvv --ffi`: Passes

### CODE REVIEW

File: programs/build.rs [hunk 1]
```diff
@@ -15,7 +15,7 @@ fn main() {
     for program in programs.clone() {
         let args = BuildArgs {
             elf_name: program.clone(),
-            output_directory: "programs/elf".to_string(),
+            output_directory: format!("programs/{}/elf", program),
```
Comment (emostov) [inline programs/build.rs:18]: Same rec as in the mono pr - I prefer to have all build artifacts in some `target` sub dir

File: programs/build.rs [hunk 2]
```diff
@@ -1,16 +1,30 @@
...
 // Paths where the generated Solidity files will be written.
 const SOLIDITY_PROGRAM_ID_PATH: &str = "../contracts/src/ProgramID.sol";
 const SOLIDITY_DEPLOY_SCRIPT_PATH: &str = "../contracts/script/Deployer.s.sol";
 
+// Add your zkVM programs here.
+const PROGRAM_NAMES: &[&str] = &["square-root"];
+
 fn main() {
-    let guests = risc0_build::embed_methods();
+    let programs: Vec<String> = PROGRAM_NAMES.to_vec().iter().map(|s| s.to_string()).collect();
```
Comment (emostov) [inline programs/build.rs:12]: you might be able to instead do 
```
let programs: Vec<_> = PROGRAM_NAMES.into_iter().map(String::to_string).collect();
```

File: programs/square-root/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,12 @@
+[package]
+name = "program"
+version = "0.1.0"
+edition = "2021"
+
+[dependencies]
+sp1-zkvm = { workspace = true, features = ["libm", "lib"], default-features = false }
+alloy = { workspace = true, features = ["sol-types"], default-features = false }
+alloy-sol-types = { workspace = true }
```
Comment (emostov) [inline programs/square-root/Cargo.toml:9]: we should be able to just use `alloy` here

File: programs/square-root/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,23 @@
+#![no_main]
+sp1_zkvm::entrypoint!(main);
+
+use alloy::primitives::U256;
+use alloy_sol_types::{sol, SolType, SolValue};
```
Comment (emostov) [inline programs/square-root/src/main.rs:5]:
```suggestion
use alloy::sol_types::{sol, SolType, SolValue};
```

Comment (emostov) [reply programs/square-root/src/main.rs:5]: I'm not positive about the `sol!` import working but SolType and SolValue should work from this

Comment (emostov) [reply programs/square-root/src/main.rs:5]: seems to work here https://github.com/InfinityVM/InfinityVM/blob/4e8a6c7b62ffd8ca6f225b12934284f76b1370f8/examples/clob/core/src/api/program.rs#L10

File: zkvm-utils/src/main.rs [hunk 5]
```diff
@@ -10,9 +10,10 @@ use alloy::{
 use alloy_sol_types::{sol, SolType};
```
Comment (emostov) [inline zkvm-utils/src/main.rs:10]:
```suggestion
use alloy::sol_types::{sol, SolType};
use alloy::hex;
```

File: zkvm-utils/src/sol.rs [hunk 6]
```diff
@@ -7,22 +7,32 @@ use std::{
 
 use alloy::primitives::hex;
 use anyhow::{anyhow, bail, Context, Result};
-use risc0_build::GuestListEntry;
-use risc0_zkp::core::digest::Digest;
+use ivm_zkvm::Zkvm;
 
 const PROGRAM_ID_LIB_HEADER: &str = r#"pragma solidity ^0.8.13;
```
Comment (emostov) [inline zkvm-utils/src/sol.rs:12]: I assume it doesn't matter if these are less then 8.28? Still works with Cancun?

**Review by emostov — COMMENTED** [review]
Looking good. Main feedback is that would be nice to just rely on program ID from builds and store all build artifact in the root target dir

File: programs/build.rs (hunk 1 above)
Comment (MaanavKhaitan) [reply programs/build.rs:18]: updated

File: programs/build.rs (hunk 2 above)
Comment (MaanavKhaitan) [reply programs/build.rs:12]: hmm i wasnt able to get this to work lol not gonna spend time on this rn

File: programs/square-root/src/main.rs (hunk 4 above)
Comment (MaanavKhaitan) [reply programs/square-root/src/main.rs:5]: fixed

File: programs/square-root/Cargo.toml (hunk 3 above)
Comment (MaanavKhaitan) [reply programs/square-root/Cargo.toml:9]: sounds good updated

File: zkvm-utils/src/main.rs [hunk 7]
```diff
@@ -63,22 +70,31 @@ enum Command {
...
+            program_id,
+            onchain_input,
+            job_id,
+            max_cycles,
+        } => {
             let job_id_decoded: [u8; 32] =
                 hex::decode(job_id.strip_prefix("0x").unwrap_or(&job_id))?.try_into().unwrap();
             execute_onchain_job_ffi(
                 guest_binary_path,
+                hex::decode(program_id.strip_prefix("0x").unwrap_or(&program_id))?,
```
[the remark this answers is not in the document]
Comment (MaanavKhaitan) [reply zkvm-utils/src/main.rs]: ah yes you're right just fixed

File: zkvm-utils/src/sol.rs [hunk 8]
```diff
@@ -65,8 +65,8 @@ pub fn generate_solidity_files(program_names: Vec<String>, opts: &Options) -> Re
         .map(|name| {
             let elf_path = format!("elf/{name}");
             let elf = std::fs::read(elf_path).unwrap();
-            let program_id = ivm_zkvm::Sp1.derive_verifying_key(&elf).unwrap();
-            let elf_path_sol = format!("programs/elf/{name}");
+            let program_id = ivm_zkvm::Sp1.derive_program_id(&elf).unwrap();
```
[the remark this answers is not in the document]
Comment (MaanavKhaitan) [reply zkvm-utils/src/sol.rs:68]: this is already happening at build time

[the remark this answers is not in the document]
Comment (emostov) [reply zkvm-utils/src/sol.rs]: this line is re-deriving it tho https://github.com/InfinityVM/InfinityVM/blob/bc5500f29cb1f8c580c63040eeccbc7aa641308d/crates/zkvm/src/lib.rs#L108

File: programs/build.rs [hunk 9]
```diff
@@ -1,16 +1,29 @@
...
+
+    // For each program, build the ELF.
+    for program in programs.clone() {
+        build_sp1_program(
+            program.as_str(),
+            program.as_str(),
+            format!("target/sp1/{program}").as_str(),
+        );
+    }
```
Comment (emostov) [inline programs/build.rs:21]: if this becomes painful in the future we can use to parallelize this https://docs.rs/rayon/latest/rayon/

Comment (MaanavKhaitan) [reply programs/build.rs:21]: ah sounds great thanks

**Pull request merged**